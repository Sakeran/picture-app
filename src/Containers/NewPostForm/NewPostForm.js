import React from 'react';
import PropTypes from 'prop-types';

import Formsy from 'formsy-react';
import TextInput from '../formComponents/TextInput';
import TextField from '../formComponents/TextField';

import youtubeRegex from 'youtube-regex';
import getYoutubeId from 'get-youtube-id';

import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

class NewPostForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      canSubmit: false,
      redirect: false
    };
  }

  enableButton = () => {
    this.setState({
      canSubmit: true
    });
  }

  disableButton = () => {
    this.setState({
      canSubmit: false
    })
  }


  submit = ({link, title, description}) => {
    console.log(link, title, description);
    this.setState({
      redirect: true
    });
  }

  render() {
    if (!this.props.user || this.state.redirect) {
      return <Redirect to="/" />
    }
    return (
      <Formsy.Form onValidSubmit={this.submit}
                   onValid={this.enableButton}
                   onInvalid={this.disableButton}>
        <TextInput name="link"
                   title="Link to an image or Youtube video:"
                   validations={{
                     isUrl: true,
                     isYoutube: (values, value) => {
                       if(!youtubeRegex().test(value)) {
                         return true; // Validation irrelevant.
                       }
                       return !!getYoutubeId(value);
                     }
                   }}
                   validationErrors={{
                     isUrl: "Link must be a valid URL",
                     isYoutube: "Invalid Youtube Link"
                   }}
                   required
                   value=""/>
        <TextInput name="title"
                   title="Give your post a title:"
                   validations="maxLength: 20"
                   validationError="Title must be 20 characters or less"
                   required
                   value="" />
        <TextField name="description"
                  title="Description (optional)"
                  value=""/>
        <input type="submit"
                className="button"
                disabled={!this.state.canSubmit}
                value="Post" />
      </Formsy.Form>
    );
  }
}

NewPostForm.propTypes = {
  sendFunc: PropTypes.func.isRequired,
  user: PropTypes.object,
  flashSuccess: PropTypes.func,
  flashError: PropTypes.func
};

const mapStateToProps = (state) => ({
  user: state.common.user
});

const mapDispatchToProps = (dispatch) => ({
  flashSuccess: (msg) => dispatch({type: 'FLASH_SUCCESS', message: msg}),
  flashError: (msg) => dispatch({type: 'FLASH_ERROR', message: msg})
});

const ConnectedNewPostForm = connect(mapStateToProps, mapDispatchToProps)(NewPostForm);

export { NewPostForm };
export default ConnectedNewPostForm;
