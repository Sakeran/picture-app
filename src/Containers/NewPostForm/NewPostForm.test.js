import React from 'react';
import { NewPostForm } from './NewPostForm';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

test('renders all fields correctly', () => {
  const fn = () => null;
  const component = mount(<NewPostForm sendFunc={fn}/>);
  expect(component.find('input[name="link"]').exists()).toBe(true);
  expect(component.find('input[name="title"]').exists()).toBe(true);
  expect(component.find('textarea[name="description"]').exists()).toBe(true);
  expect(component.find('input[type="submit"]').exists()).toBe(true);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
