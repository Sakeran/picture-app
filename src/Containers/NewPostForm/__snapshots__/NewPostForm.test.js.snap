// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all fields correctly 1`] = `
<NewPostForm
  sendFunc={[Function]}
>
  <Formsy
    onChange={[Function]}
    onError={[Function]}
    onInvalid={[Function]}
    onInvalidSubmit={[Function]}
    onSubmit={[Function]}
    onSuccess={[Function]}
    onValid={[Function]}
    onValidSubmit={[Function]}
    preventExternalInvalidation={false}
    validationErrors={null}
  >
    <form
      onSubmit={[Function]}
    >
      <Formsy(TextInput)
        name="link"
        required={true}
        title="Link to an image or Youtube video:"
        validationError=""
        validationErrors={
          Object {
            "isUrl": "Link must be a valid URL",
            "isYoutube": "Invalid Youtube Link",
          }
        }
        validations={
          Object {
            "isUrl": true,
            "isYoutube": [Function],
          }
        }
        value=""
      >
        <TextInput
          getErrorMessage={[Function]}
          getErrorMessages={[Function]}
          getValue={[Function]}
          hasValue={[Function]}
          isFormDisabled={[Function]}
          isFormSubmitted={[Function]}
          isPristine={[Function]}
          isRequired={[Function]}
          isValid={[Function]}
          isValidValue={[Function]}
          name="link"
          required={true}
          resetValue={[Function]}
          setValidations={[Function]}
          setValue={[Function]}
          showError={[Function]}
          showRequired={[Function]}
          title="Link to an image or Youtube video:"
          validationError=""
          validationErrors={
            Object {
              "isUrl": "Link must be a valid URL",
              "isYoutube": "Invalid Youtube Link",
            }
          }
          validations={
            Object {
              "isUrl": true,
              "isYoutube": [Function],
            }
          }
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="link"
            >
              Link to an image or Youtube video:
            </label>
            <input
              name="link"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="field-error"
            />
          </div>
        </TextInput>
      </Formsy(TextInput)>
      <Formsy(TextInput)
        name="title"
        required={true}
        title="Give your post a title:"
        validationError="Title must be 20 characters or less"
        validationErrors={Object {}}
        validations="maxLength: 20"
        value=""
      >
        <TextInput
          getErrorMessage={[Function]}
          getErrorMessages={[Function]}
          getValue={[Function]}
          hasValue={[Function]}
          isFormDisabled={[Function]}
          isFormSubmitted={[Function]}
          isPristine={[Function]}
          isRequired={[Function]}
          isValid={[Function]}
          isValidValue={[Function]}
          name="title"
          required={true}
          resetValue={[Function]}
          setValidations={[Function]}
          setValue={[Function]}
          showError={[Function]}
          showRequired={[Function]}
          title="Give your post a title:"
          validationError="Title must be 20 characters or less"
          validationErrors={Object {}}
          validations="maxLength: 20"
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="title"
            >
              Give your post a title:
            </label>
            <input
              name="title"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="field-error"
            />
          </div>
        </TextInput>
      </Formsy(TextInput)>
      <Formsy(TextField)
        name="description"
        title="Description (optional)"
        validationError=""
        validationErrors={Object {}}
        value=""
      >
        <TextField
          getErrorMessage={[Function]}
          getErrorMessages={[Function]}
          getValue={[Function]}
          hasValue={[Function]}
          isFormDisabled={[Function]}
          isFormSubmitted={[Function]}
          isPristine={[Function]}
          isRequired={[Function]}
          isValid={[Function]}
          isValidValue={[Function]}
          name="description"
          resetValue={[Function]}
          setValidations={[Function]}
          setValue={[Function]}
          showError={[Function]}
          showRequired={[Function]}
          title="Description (optional)"
          validationError=""
          validationErrors={Object {}}
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="description"
            >
              Description (optional)
            </label>
            <textarea
              name="description"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="field-error"
            />
          </div>
        </TextField>
      </Formsy(TextField)>
      <input
        className="button"
        disabled={true}
        type="submit"
        value="Post"
      />
    </form>
  </Formsy>
</NewPostForm>
`;
